Listen for GitLab job_event webhooks

sgallagher opened the pull request:
Listens for the events and queries the GitLab API for additional
information leading to the logs of failed builds.

Retrieves the logs and analyzes them to determine the correct log to
submit to the Log Detective processing.

Fixes: https://issues.redhat.com/browse/LD-8

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

sgallagher commented:
Still to do:

- Decide how we're going to maintain the opt-in list of packages and implement that
- Rework the tempfile implementation. It was originally done with a tempdir when I was doing a full zipfile extraction, but since we're now using the `zipfile` module, it can be simplified.
- Determine if we want to be maintaining the log content entirely in-memory or if we can pass a file-like object into Log Detective as a stream from the `zipfile` interface. If so, revise the tempfile to stay available until full processing is complete.

TomasTomecek commented:
We need to improve that linter because it's impossible to code review right now.

jpodivin commented:
> > Determine if we want to be maintaining the log content entirely in-memory or if we can pass a > file-like object into Log Detective as a stream from the zipfile interface. If so, revise the > tempfile to stay available until full processing is complete.
> 
> This is interesting.
> 
> It's true that some logs can be colossal and could crash our service when loaded all in memory. I'm curious what would happen in openshift for such a case.

What's the largest log that can, under reasonable conditions, come out of rpm build? I can't imagine it getting past 1GB.

sgallagher commented:
> What's the largest log that can, under reasonable conditions, come out of rpm build? I can't imagine it getting past 1GB.

The largest example I can find recently would be for [gcc](https://kojipkgs.fedoraproject.org//packages/gcc/15.0.1/0.10.fc43/data/logs/x86_64/build.log), which is around 135MiB of build.log. That's probably about the maximum size under "reasonable conditions". But it occurs to me that it's also not impossible for a log to end up filled with spew (such as getting caught in a tight loop that includes a log message). I see us as having two options in that case: 1) check the content-length and just refuse to process logs over a certain size or 2) pass the logs as a stream so that Log Detective can handle chunks of it at a time.

If Log Detective needs the whole thing to be in-memory, that probably limits us to option 1.

TomasTomecek commented:
the content length header is an excellent solution!

TomasTomecek commented:
Please rebase since the ruff changes are merged now.

TomasTomecek commented on logdetective/server/server.py:
```diff
@@ -260,3 +273,167 @@ async def analyze_log_stream(build_log: BuildLog):
+    lists the full specification for the messages sent for job events."""
+
+    # Handle the message in the background so we can return 200 immediately
+    background_tasks.add_task(process_gitlab_job_event, job_hook)
```
still fascinating to me this background tasks functionality is built into a web framework

I hope we'll be able to decode/monitor based on logs what's happening inside

TomasTomecek replied:
Excellent point, I'll create a task for this.

TomasTomecek commented on logdetective/server/server.py:
```diff
@@ -260,3 +273,167 @@ async def analyze_log_stream(build_log: BuildLog):
+    merge_request_id = int(match.group(1))
+
+    # Retrieve the build logs from the merge request artifacts and preprocess them
+    preprocessed_log = await retrieve_and_preprocess_koji_logs(job)
+
+    # Submit log to Log Detective and await the results.
+    response = await submit_log_to_llm(preprocessed_log)
+    preprocessed_log.close()
+
+    # Add the Log Detective response as a comment to the merge request
+    await comment_on_mr(merge_request_id, response)
```
nice and clear :)

TomasTomecek commented on logdetective/server/server.py:
```diff
@@ -260,3 +273,167 @@ async def analyze_log_stream(build_log: BuildLog):
+            failed_arch = "ppc64le"
+        elif "s390x" in failed_arches:
+            failed_arch = "s390x"
+        else:
```
I wonder how will this code react to SRPM build failures

sgallagher replied:
I didn't find an example to test, so I probably need to create one...

I'm pretty sure that it still returns an archful log though. Even the noarch builds for python packages, et. al. do that for binary RPMs.

TomasTomecek replied:
Interesting, every buildsystem behaves differently.

If it doesn't work, we'll find out soon enough :)

TomasTomecek commented on server/config.yml:
```diff
@@ -8,3 +8,6 @@ extractor:
   api_token: glpat-XXXXXX
+general:
+  packages:
+    - osci-internal-test-package
```
perfect, there is a package we can play with \o/

TomasTomecek reviewed (approved):
Love all the comments in the code and how readable it is, very nice work!!

sgallagher commented:
> the content length header is an excellent solution!

It looks like we actually can stream-process it, which is far more memory-efficient, but I think I'll still implement the header check and limit us to a (configurable) 300 MiB log file. If it's bigger than that, odds are good that the log isn't actually all that useful.


> Please rebase since the ruff changes are merged now.

Done

sgallagher commented:
I added comments indicating that some of the functions are placeholders that will be implemented in future merge requests. I also added a check for the size of `artifacts.zip` so we will just raise an error if we are being asked to handle very large files. The maximum size is configurable.

TomasTomecek reviewed (approved):
Amazing! Let's deploy and test it out :)
